### PR TITLE
Communities Join Screens - Implement all permutations of Context Drawer options 

### LINKED ISSUE #14464: Communities Join Screens - Implement all permutations of Context Drawer options
The communities join screens context drawers has lots of possible states dependent on different values, e.g joined, not-joined, token-gated, not token-gated etc

See the designs here:
https://www.figma.com/file/apCsJHT8J58JXbJD2igtbd/Context-Drawers-for-Mobile?node-id=167%3A107129&t=aRVvvzqko8zHGE59-0

Implement the various combinations for each state.
Discuss with design team before beginning this work.

Tests:
Add jest tests covering the function which filters the correct bottom sheet options and ensures the correct values are shown dependent on given values.

### PR DESCRIPTION
**Pull request by J-Son89**
fixes: https://github.com/status-im/status-mobile/issues/14464

There are three follow up issues:

corresponding actions will be implemented in a follow up issue. 
denied request options -  as I was not receiving the message from the backend so I need to look into it more.
handle mute community 

TO TEST:
 Go to a community overview page
 
 Click the three dot options to see the bottom sheet.
 
 The options should be different depending on whether the user is joined etc 
 
 This pr does not implement the functionality but just the designs.
 
 
See the designs: https://www.figma.com/file/apCsJHT8J58JXbJD2igtbd/Context-Drawers-for-Mobile?node-id=167%3A107129&t=aRVvvzqko8zHGE59-0

### CODE REVIEW

File: src/status_im2/contexts/communities/menus/community_options/__tests__/community_options_component_spec.cljs
Comment (J-Son89) [inline src/status_im2/contexts/communities/menus/community_options/__tests__/community_options_component_spec.cljs]: @ilmotta WDYT of this for tests in status_im2?

File: src/status_im2/contexts/communities/menus/community_options/view.cljs
Comment (ilmotta) [inline src/status_im2/contexts/communities/menus/community_options/view.cljs]: https://guide.clojure.style/#terminate-files-with-a-newline

While this is not yet configured by the formatter, please, configure your own text editor to auto terminate files with new lines.

File: src/status_im2/contexts/communities/menus/community_options/view.cljs
Comment (ilmotta) [inline src/status_im2/contexts/communities/menus/community_options/view.cljs]: https://guide.clojure.style/#one-semicolon-for-margin-comments

File: src/status_im2/contexts/communities/menus/community_options/view.cljs
Comment (ilmotta) [inline src/status_im2/contexts/communities/menus/community_options/view.cljs:48]: Missing translation?

File: src/status_im2/contexts/communities/menus/community_options/view.cljs
Comment (ilmotta) [inline src/status_im2/contexts/communities/menus/community_options/view.cljs:119]: Extra line

File: src/status_im2/contexts/communities/menus/community_options/view.cljs
Comment (ilmotta) [inline src/status_im2/contexts/communities/menus/community_options/view.cljs:137]: @J-Son89 While I think you did the right thing in this PR to get something working, if you go on and talk to some folks in the Desktop Team, they'll tell you that we should, ideally, move important business rules to the backend, so that we reduce the chance of Status clients drifting.

So although I'm definitely not against how this PR handles all the different permutations, I like to think this is a temporary solution, because I think status-go should be responsible for saying which capabilities/features are available clients.

The same kind of feedback was given by someone from the Desktop Team in relation to the Activity Center. I'm sure there are many other instances of business rules that should be moved to the backend, so this is not a new problem for sure.

What should we do? I'm not sure right now, given the priorities to get dogfooding working, but something to keep in mind for future changes.

File: src/status_im2/contexts/communities/menus/community_options/view.cljs
Comment (ilmotta) [inline src/status_im2/contexts/communities/menus/community_options/view.cljs:159]: I would definitely prefer if all the logic to decide which actions are available could be returned by a subscription and the UI would just dynamically dispatch to different components.

Let me know if you're interested in pursuing this improvement, and I can share more details.

File: src/status_im2/contexts/communities/menus/leave/view.cljs
Comment (ilmotta) [inline src/status_im2/contexts/communities/menus/leave/view.cljs:12]: Extra line

File: src/status_im2/contexts/communities/menus/leave/view.cljs
Comment (ilmotta) [inline src/status_im2/contexts/communities/menus/leave/view.cljs:23]: Zombie code lying around? Why not a GH issue and delete the TODO and zombie code?

File: src/status_im2/contexts/communities/menus/request_to_join/view.cljs
Comment (ilmotta) [inline src/status_im2/contexts/communities/menus/request_to_join/view.cljs]: Extra line

File: src/quo2/components/messages/gap.cljs
Comment (ilmotta) [inline src/quo2/components/messages/gap.cljs]: Why do we need to use a different resource if the component is running under the Jest umbrella? Have you tried using `with-redefs` in tests that need different resources, instead of changing the production code?

File: src/status_im/ui/components/icons/icons.clj
Comment (ilmotta) [inline src/status_im/ui/components/icons/icons.clj]: Same comment I made about the check against `js/global.__TEST__`, why do we need a different resource if the code is running via Jest? And have you tried using `with-redefs` in tests?

There's likely a better way, but maybe you've tried something already?

File: src/status_im2/contexts/communities/menus/community_options/__tests__/community_options_component_spec.cljs
Comment (ilmotta) [inline src/status_im2/contexts/communities/menus/community_options/__tests__/community_options_component_spec.cljs]: We can use some helpers from the namespace `test-helpers.component`. In particular, the macros `test-helpers.component/describe` and `test-helpers.component/test` can help with test readability by grouping related tests, instead of doing what you did here to start the test descriptions with a common prefix, like `joined options - `.

File: src/status_im2/contexts/communities/menus/community_options/__tests__/community_options_component_spec.cljs
Comment (ilmotta) [inline src/status_im2/contexts/communities/menus/community_options/__tests__/community_options_component_spec.cljs]: From the guidelines and previous discussions, if the component test is isolated inside a single component directory, then we should name the file just `menus/community_options/component_spec.cljs`.

https://github.com/status-im/status-mobile/blob/8d08de3d4b736cbd1948ca4add07bc927aba8378/doc/new-guidelines.md#test-structure

File: src/status_im2/contexts/communities/menus/community_options/__tests__/community_options_component_spec.cljs
Comment (ilmotta) [inline src/status_im2/contexts/communities/menus/community_options/__tests__/community_options_component_spec.cljs]: Extra line

File: src/status_im2/contexts/communities/menus/community_options/__tests__/community_options_component_spec.cljs
Comment (ilmotta) [inline src/status_im2/contexts/communities/menus/community_options/__tests__/community_options_component_spec.cljs]: Sorry, another comment about extra line.

File: src/status_im2/contexts/communities/menus/community_options/__tests__/community_options_component_spec.cljs
Comment (ilmotta) [inline src/status_im2/contexts/communities/menus/community_options/__tests__/community_options_component_spec.cljs]: Dangling new lines.

File: src/status_im2/contexts/communities/menus/community_options/__tests__/community_options_component_spec.cljs
Comment (ilmotta) [reply src/status_im2/contexts/communities/menus/community_options/__tests__/community_options_component_spec.cljs]: @J-Son89 I'm afraid Jest is not automatically cleaning up registered subscriptions in re-frame's global registry. This is not a problem right now because we don't have many tests competing for the same subscriptions, but I believe this solution won't work soon, because subscriptions registered during tests will have side-effects in other tests. At the very least I would expect some tests to "randomly" fail due to one test having registered a subscription that affects another one. 

Re-frame's registry is global, so one of the main features of the library https://github.com/day8/re-frame-test is to help clean up state between tests. You know we use this lib for writing integration tests and some event handler tests.

We either use some of the features provided by that lib to clean up the registry or we do it ourselves, like I did with the macro `status-im.test-helpers/deftest-sub`.

---

I think what you wrote here is fine for now, but someone will eventually be affected by the problem I described and we'll need to find a reliable and easy solution to setup subscriptions and clean them up between component tests.

File: src/status_im2/contexts/communities/menus/community_options/__tests__/community_options_component_spec.cljs
Comment (ilmotta) [inline src/status_im2/contexts/communities/menus/community_options/__tests__/community_options_component_spec.cljs]: Excellent to see you bringing component tests outside the quo library!

File: src/status_im2/contexts/communities/menus/community_options/view.cljs
Comment (J-Son89) [reply src/status_im2/contexts/communities/menus/community_options/view.cljs:48]: I have a follow up issue to handle the muted/unmuted properly

File: src/status_im2/contexts/communities/menus/community_options/view.cljs
Comment (J-Son89) [reply src/status_im2/contexts/communities/menus/community_options/view.cljs:137]: yeah that sounds fair, we can work to get this in and create a follow up task to communicate with the Desktop team so we can get this logic from their side. CC @cammellos - who can I can discuss with about topics like this?

File: src/status_im2/contexts/communities/menus/community_options/view.cljs
Comment (J-Son89) [reply src/status_im2/contexts/communities/menus/community_options/view.cljs:159]: yeah for sure, we can do this 👍

File: src/status_im2/contexts/communities/menus/leave/view.cljs
Comment (J-Son89) [reply src/status_im2/contexts/communities/menus/leave/view.cljs:23]: sure, I'll create a task on this

File: src/quo2/components/messages/gap.cljs
Comment (J-Son89) [reply src/quo2/components/messages/gap.cljs]: hmm, it's not the test directly that needs the resources but some file that\s a dependency of the test file. I think we just ultimately need to configure something and it should resolve this.

File: src/status_im2/contexts/communities/menus/community_options/__tests__/community_options_component_spec.cljs
Comment (J-Son89) [reply src/status_im2/contexts/communities/menus/community_options/__tests__/community_options_component_spec.cljs]: yep, will update this file to the new helpers!

File: src/status_im/ui/components/icons/icons.clj
Comment (J-Son89) [reply src/status_im/ui/components/icons/icons.clj]: oh, in reality this file will be removed once we remove all uses of the old icons. 

This code exists in the quo2/icons/icon.clj too, ideally we can configure this correctly so we don't need this logic within files but in the configuration files instead.

File: src/status_im2/contexts/communities/menus/community_options/view.cljs
Comment (cammellos) [reply src/status_im2/contexts/communities/menus/community_options/view.cljs:137]: Generally status-go does not advertise capabilities, it offers the API, but we leave responsibility to the client to know which capabilities it offers, since the client is fully in control of which version to use (we could of course change/discuss if we think it's valuable, but I don't think there's any precedence for this in the codebase at the moment, though I might be a bit behind development after vacation :) ).

Permissions are checked in status-go, for example in a community, we manage that in `status-go`, `canPost?` comes to mind, but it's a permission check, rather than a capabilities check.

Though maybe that's something discussed elsewhere @ilmotta would you mind pointing me to the conversation you had please?

Comment (ilmotta) [reply src/status_im2/contexts/communities/menus/community_options/view.cljs:137]: Thanks @cammellos for sharing the current strategy.

The conversation was private, but it was related to the Activity Center and if `status-go` should be responsible or not for telling if a notification requires a call to action or not. Currently the desktop and mobile decide on the front because it's much more convenient. That's why I made the comment in this PR, because it looks like the kind of logic that could be shared between clients.

But I think we haven't necessarily agreed on what a capability really is in the first place, so maybe this word is bringing more confusion to the discussion.

I was under the impression we wanted to follow a more "heavy server, light client" kind of strategy, so that we reduce the chances of clients drifting (when they use the same status-go version of course).

Maybe I just got things wrong, but it might be beneficial to bring this topic to some people in the desktop team as well. We might all have different perspectives on what should reside on the backend vs frontend.

File: src/status_im2/contexts/communities/menus/leave/view.cljs
Comment (J-Son89) [reply src/status_im2/contexts/communities/menus/leave/view.cljs:23]: tbh, I prefer to leave the code as the component is correct so it holds the structure which is important.

Comment (J-Son89) [reply src/status_im2/contexts/communities/menus/leave/view.cljs:23]: I will make a quick follow up and address this though.

File: src/status_im2/contexts/communities/menus/community_options/view.cljs
Comment (cammellos) [reply src/status_im2/contexts/communities/menus/community_options/view.cljs:137]: Sure,
feel free to set up a conversation with the relevant people, I'll be happy to join!

File: src/status_im/ui/components/icons/icons.clj
Comment (ilmotta) [reply src/status_im/ui/components/icons/icons.clj]: I kind of understand, but which configuration files you mean would be ideal for setting up this?

Comment (J-Son89) [reply src/status_im/ui/components/icons/icons.clj]: That's what I (we) need to figure out :)

Comment (J-Son89) [reply src/status_im/ui/components/icons/icons.clj]: it's some asset loader, maybe it's in jest

File: src/quo2/components/messages/gap.cljs
Comment (ilmotta) [reply src/quo2/components/messages/gap.cljs]: > it's not the test directly that needs the resources but some file that\s a dependency of the test file.

I still don't understand :man_shrugging: Which file are you referring to that requires this path change when running via Jest? I'm just trying to understand the root cause, because it's not apparent from the diff alone.

File: src/status_im/ui/components/icons/icons.clj
Comment (ilmotta) [reply src/status_im/ui/components/icons/icons.clj]: > That's what I (we) need to figure out :)

Okay, I see now. Do we have an issue to track this problem? I would be interested to help out when I finish some stuff this week.

File: src/status_im2/contexts/communities/menus/community_options/view.cljs
Comment (J-Son89) [reply src/status_im2/contexts/communities/menus/community_options/view.cljs:137]: add me too @ilmotta 👍

**Comment by J-Son89** [comment]
@pavloburykh, @VladimrLitvinenko is it possible to get this checked? I have testing steps in the description. let me know if you need more details. :)

File: test/jest/jest.config.js
Comment (J-Son89) [inline test/jest/jest.config.js:24]: needed this for new file structure format to be detected. @ilmotta

File: test/jest/jest.config.js
Comment (J-Son89) [inline test/jest/jest.config.js:10]: this is the img  resource stub @ilmotta

File: src/status_im/ui/components/icons/icons.clj
Comment (ilmotta) [inline src/status_im/ui/components/icons/icons.clj]: This `str` call concatenates to a path starting with 3 dots, is this really correct?

File: test/jest/jest.config.js
Comment (briansztamfater) [reply test/jest/jest.config.js:10]: I think this fixes component tests failing in current `develop` branch, can we merge it first on a separate PR if this one takes too long to get merged? 

el-gato-con-botas-1551289404

Comment (ilmotta) [reply test/jest/jest.config.js:10]: @briansztamfater, do you how did we end up in the situation where `develop` has failing unit tests? I thought it wasn't possible. If the tests are deterministically failing, then the CI should've blocked the merge.

Any idea?

Comment (ilmotta) [reply test/jest/jest.config.js:10]: > [...] can we merge it first on a separate PR if this one takes too long to get merged?

In my experience, when a development branch has failing tests, the collaborator who broke the tests (and who probably has more context about how to fix them) go ahead and fix them ASAP, or the persons with most knowledge of the failure. And this should be done in a separate PR to unblock other collaborators and not affect PRs with business as usual work.

So I agree with you about trying to fix `develop` as quickly as possible, but I think such fixes should always be done on separate PRs, particularly to avoid merging things in a hurry.

Comment (briansztamfater) [reply test/jest/jest.config.js:10]: For some reason many component tests are failing locally for me (by running `make component-test`), also @cammellos was experiencing this I think. I cannot say why in CI environment is working or maybe jest tests are not being checked in the CI job?

**Comment by pavloburykh** [comment]
> @pavloburykh, @VladimrLitvinenko is it possible to get this checked? I have testing steps in the description. let me know if you need more details. :)

hey @J-Son89 ! Sure will take PR into work as soon as it will be ready for manual QA. 

Currently I see that all PR builds have failed including build with integration tests. Also, PR has not passed developers review yet. 

Just a quick reminder about our PR workflow: as soon as PR is ready for testing by manual qa, it should be moved by dev to E2E column. 

PR is ready for testing when matches following conditions:
1. It has been reviewed by dev team and has got at least one approval. Changes requests should be handled too.
2. There are no conflicts and PR is rebased to the latest develop
3. Integration tests have passed and all necessary builds have no failures. 

All detailed info about PR workflow can be found [here](https://notes.status.im/M73LlV1gT3OJ8kdSMY-ZQQ?view#When-is-a-PR-considered-to-be-Ready-for-testing-by-QA-team)

File: test/jest/jest.config.js
Comment (briansztamfater) [reply test/jest/jest.config.js:10]: > > [...] can we merge it first on a separate PR if this one takes too long to get merged?
> 
> In my experience, when a development branch has failing tests, the collaborator who broke the tests (and who probably has more context about how to fix them) go ahead and fix them ASAP, or the persons with most knowledge of the failure. And this should be done in a separate PR to unblock other collaborators and not affect PRs with business as usual work.
> 
> So I agree with you about trying to fix `develop` as quickly as possible, but I think such fixes should always be done on separate PRs, particularly to avoid merging things in a hurry.

Yeah agree, I was about to prepare a PR with this fix but just saw that we already had fixes in this one which probably are better tested

Comment (J-Son89) [reply test/jest/jest.config.js:10]: Yeah @briansztamfater @ilmotta component tests are not being run in ci yet.
@siddarthkay is working on this afaik.

As to why component tests are failing - 🤷‍♂️ but maybe make sure you only have one active terminal, e.g make run-metro is not running in another.

Feel free to ping me and send the error on and I can take a look.

Tests are passing for me in this branch anyway

File: src/status_im/ui/components/icons/icons.clj
Comment (J-Son89) [reply src/status_im/ui/components/icons/icons.clj]: nope, there should be no changes to this file. will revert 👍

**Review by ilmotta — APPROVED** [review]

File: test/jest/jest.config.js
Comment (yqrashawn) [reply test/jest/jest.config.js:10]: ~I believe it's caused by PR here https://github.com/status-im/status-mobile/pull/14735#issuecomment-1376799083~
~interface of `:toasts/upsert` changed in this pr and not all the usage of it updated~
~I'll fix it asap~

Comment (yqrashawn) [reply test/jest/jest.config.js:10]: Oops, it's not caused by this, the usage is already updated

**Comment by J-Son89** [comment]
@pavloburykh apologies, I overlooked those details. I had so many comments I assumed it was dev approved (which it is now :) ) 
Everything seems to be working fine in the CI, albeit e2e tests were broken from some changes that were merged into develop and the ios CI job is having some issues, also in other prs.

**Comment by pavloburykh** [comment]
Hey @J-Son89 Are any additional commits about to be pushed or PR is ready for testing? I see that you have just pushed some new commits.

**Comment by J-Son89** [comment]
@pavloburykh nope, I was just rebasing. code hasn't changed :)

**Comment by pavloburykh** [comment]
> @pavloburykh nope, I was just rebasing. code hasn't changed :)

cool, thanx!

**Comment by pavloburykh** [comment]
Hi @J-Son89 Got one question: do I understand correctly that currently there are no devision for communities on token gated and non token gated?

I am just in process of testing and I see that options for not joined communities are displayed with View gating option
Context Drawers for Mobile – Figma 2023-01-12 13-04-17

at the same time, after changing status to Pending or Joined for such communities, View community rules option is displayed instead of View token gating

Context Drawers for Mobile – Figma 2023-01-12 13-06-38

So I am a little bit confused about what kind of options I should expect for current types communities that we have. Cause the devision for gated/non gated which is reflected on [designs](https://www.figma.com/file/apCsJHT8J58JXbJD2igtbd/Context-Drawers-for-Mobile?node-id=791%3A275811&t=3bQR3h9Q6nPCOfEM-0) seems to be irrelevant for current app version.

**Comment by J-Son89** [comment]
> Member

Hi @pavloburykh, good spot. that was actually a bug in the code that I had it toggled on as `token-gated` when it shouldn't be. 

I have adjusted it now and will push the changes. It was only for the not joined community, everything else should have been correct.

**Comment by pavloburykh** [comment]
@J-Son89 please, rebase the PR to the latest develop to get these changes https://github.com/status-im/status-mobile/pull/14759 Cause IOS builds are failing.

**Comment by J-Son89** [comment]
> @J-Son89 please, rebase the PR to the latest develop to get these changes https://github.com/status-im/status-mobile/pull/14759 Cause IOS builds are failing. 

Sure, pushing now :)

**Comment by pavloburykh** [comment]
@J-Son89 thanx for the fixes. Ready to be merged.

**Pull request merged**
